src/main/java/org/generationcp/middleware/dao/dms/EnvironmentDao.java

```diff
@@ -315,69 +277,60 @@ public TrialEnvironments getEnvironmentsForTraits(final List<Integer> traitIds,
 			}
 
 		} catch (final HibernateException e) {
-			final String errorMessage = "Error at getEnvironmentForTraits at GeolocationDao: " + e.getMessage();
-			GeolocationDao.LOG.error(errorMessage, e);
+			final String errorMessage = "Error at getEnvironmentForTraits at EnvironmentDao: " + e.getMessage();
+			EnvironmentDao.LOG.error(errorMessage, e);
 			throw new MiddlewareQueryException(errorMessage, e);
 		}
 		return environments;
 	}
 
-	@SuppressWarnings("unchecked")
-	public Integer getLocationIdByProjectNameAndDescriptionAndProgramUUID(
-		final String projectName,
-		final String locationDescription, final String programUUID) {
-		try {
-			final String sql = "SELECT DISTINCT e.nd_geolocation_id"
-				+ " FROM nd_experiment e"
-				+ " INNER JOIN nd_geolocation g ON g.nd_geolocation_id = e.nd_geolocation_id"
-				+ " INNER JOIN project p ON e.project_id = p.project_id "
-				+ " INNER JOIN project st ON st.project_id = p.study_id "
-				+ " WHERE st.name = :projectName"
-				+ "   and st.program_uuid = :programUUID" + "   and g.description = :locationDescription";
-			final Query query = this.getSession().createSQLQuery(sql);
-			query.setParameter("projectName", projectName);
-			query.setParameter("locationDescription", locationDescription);
-			query.setParameter("programUUID", programUUID);
-			final List<Integer> list = query.list();
-			if (list != null && !list.isEmpty()) {
-				return list.get(0);
-			}
+	public Integer getNextInstanceNumber(final Integer datasetId) {
+		final StringBuilder sb = new StringBuilder();
+		sb.append("SELECT max(observation_unit_no) ");
+		sb.append(" FROM nd_experiment e ");
+		sb.append(" WHERE e.project_id = :datasetId and e.type_id = 1020" );
```

Comment:
@darla-leafnode I have seen in many places that we are hardcoding some ids, can it be possible to use TermId instead?